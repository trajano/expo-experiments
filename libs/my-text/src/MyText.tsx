import { createContext, forwardRef, useContext, useMemo } from 'react';
import { StyleSheet, Text, type TextProps, type TextStyle } from 'react-native';
import { mapToEmbeddedFontFamily, mapToGoogleFontKey } from './mapToGoogleFontKey';
import type { FontFamilyMapper, LineHeightMapper } from './MyText.types';

/**
 * FontStyleContext is used to define the shape of the font styles
 * that will be tracked and inherited by nested MyText components.
 */
interface FontStyleContext {
    fontFamily?: string;
    fontStyle?: TextStyle['fontStyle'];
    fontWeight?: TextStyle['fontWeight'];
    mapToFontFamily: FontFamilyMapper;
    mapToLineHeight: LineHeightMapper;
}

/**
 * MyTextContext is a React Context that provides the font styles to
 * all nested MyText components.
 */
const MyTextContext = createContext<FontStyleContext>({
    mapToFontFamily: mapToGoogleFontKey,
    mapToLineHeight: () => undefined,
});

export type MyTextProps = TextProps & {
    /**
     * An optional mapper that would get the final font family given the logical font family, font weight and font style.
     * This defalts to {@link mapToGoogleFontKey}
     */
    mapToFontFamily?: FontFamilyMapper,
    /**
     * An optional mapper that would take the fontSize and translate it to a lineHeight value to ensure consistency across platforms.
     * This defalts to always return `undefined` which lets React Native handle it natively.
     */
    mapToLineHeight?: LineHeightMapper,
}
/**
 * MyText is a custom Text component that inherits font styles
 * (fontFamily, fontStyle, fontWeight) from its parent MyText components
 * unless explicitly overridden.  It presumes the use of Expo Google Fonts  https://docs.expo.dev/develop/user-interface/fonts/#use-google-fonts 
 * and are loaded using useExpoGoogleFonts hook
 *
 * @param props - Props passed to the Text component.
 * @param ref - Ref forwarded to the Text component.
 * @returns A Text component with inherited font styles.
 */
export const MyText = forwardRef<Text, MyTextProps>(({ children, mapToFontFamily, mapToLineHeight, style, ...props }, ref) => {
    // Retrieve the parent styles from context
    const parentStyles = useContext(MyTextContext);

    // Flatten the incoming style to simplify processing
    const flattenedStyle = useMemo(() => StyleSheet.flatten(style) ?? {}, [style]);

    // Use a single useMemo to compute both currentFontStyles and combinedStyle
    const { currentFontStyles, combinedStyle } = useMemo(() => {
        const mapperToFontFamily = mapToFontFamily ?? mapToGoogleFontKey;
        const mapperToLineHeight = mapToLineHeight ?? (() => undefined);
        const currentFontStyles: FontStyleContext = {
            fontFamily: flattenedStyle.fontFamily ?? parentStyles.fontFamily,
            fontStyle: flattenedStyle.fontStyle ?? parentStyles.fontStyle,
            fontWeight: flattenedStyle.fontWeight ?? parentStyles.fontWeight,
            mapToFontFamily: mapperToFontFamily,
            mapToLineHeight: mapperToLineHeight
        };

        const mappedFontFamily = mapperToFontFamily(
            currentFontStyles.fontFamily,
            currentFontStyles.fontWeight,
            currentFontStyles.fontStyle,
        );

        const mappedLineHeight = flattenedStyle.lineHeight ?? mapperToLineHeight(flattenedStyle.fontSize);

        const combinedStyle: TextStyle = {
            ...flattenedStyle,
            fontFamily: mappedFontFamily ?? currentFontStyles.fontFamily, // Fallback to original fontFamily if mapping fails
            fontStyle: mappedFontFamily ? undefined : currentFontStyles.fontStyle, // Fallback to original fontStyle if mapping fails
            fontWeight: mappedFontFamily ? undefined : currentFontStyles.fontWeight, // Fallback to original fontWeight if mapping fails
            lineHeight: mappedLineHeight,
        };

        return { currentFontStyles, combinedStyle };
    }, [flattenedStyle, parentStyles]);

    return (
        <MyTextContext.Provider value={currentFontStyles}>
            <Text ref={ref} style={combinedStyle} {...props}>
                {children}
            </Text>
        </MyTextContext.Provider>
    );
});

/**
 * MyText is a custom Text component that inherits font styles
 * (fontFamily, fontStyle, fontWeight) from its parent MyText components
 * unless explicitly overridden.  It presumes the use of Expo Google Fonts  https://docs.expo.dev/develop/user-interface/fonts/#use-google-fonts 
 * and are loaded using useExpoGoogleFonts hook
 *
 * @param props - Props passed to the Text component.
 * @param ref - Ref forwarded to the Text component.
 * @returns A Text component with inherited font styles.
 */
export const MyTextE = forwardRef<Text, MyTextProps>(({ children, mapToFontFamily, mapToLineHeight, style, ...props }, ref) => {
    // Retrieve the parent styles from context
    const parentStyles = useContext(MyTextContext);

    // Flatten the incoming style to simplify processing
    const flattenedStyle = useMemo(() => StyleSheet.flatten(style) ?? {}, [style]);

    // Use a single useMemo to compute both currentFontStyles and combinedStyle
    const { currentFontStyles, combinedStyle } = useMemo(() => {
        const mapperToFontFamily = mapToFontFamily ?? mapToEmbeddedFontFamily;
        const mapperToLineHeight = mapToLineHeight ?? (() => undefined);
        const currentFontStyles: FontStyleContext = {
            fontFamily: flattenedStyle.fontFamily ?? parentStyles.fontFamily,
            fontStyle: flattenedStyle.fontStyle ?? parentStyles.fontStyle,
            fontWeight: flattenedStyle.fontWeight ?? parentStyles.fontWeight,
            mapToFontFamily: mapperToFontFamily,
            mapToLineHeight: mapperToLineHeight
        };

        const mappedFontFamily = mapperToFontFamily(
            currentFontStyles.fontFamily,
            currentFontStyles.fontWeight,
            currentFontStyles.fontStyle,
        );

        const mappedLineHeight = flattenedStyle.lineHeight ?? mapperToLineHeight(flattenedStyle.fontSize);

        const combinedStyle: TextStyle = {
            ...flattenedStyle,
            fontFamily: mappedFontFamily ?? currentFontStyles.fontFamily, // Fallback to original fontFamily if mapping fails
            fontStyle: mappedFontFamily ? undefined : currentFontStyles.fontStyle, // Fallback to original fontStyle if mapping fails
            fontWeight: mappedFontFamily ? undefined : currentFontStyles.fontWeight, // Fallback to original fontWeight if mapping fails
            lineHeight: mappedLineHeight,
        };

        return { currentFontStyles, combinedStyle };
    }, [flattenedStyle, parentStyles]);

    return (
        <MyTextContext.Provider value={currentFontStyles}>
            <Text ref={ref} style={combinedStyle} {...props}>
                {children}
            </Text>
        </MyTextContext.Provider>
    );
});
