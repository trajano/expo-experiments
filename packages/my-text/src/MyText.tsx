import { createContext, forwardRef, useContext } from 'react';
import { Text, type TextProps } from 'react-native';
import {
  mapToEmbeddedFontFamily,
  mapToGoogleFontKey,
} from './mapToEmbeddedFontFamily';
import type {
  FontFamilyMapper,
  FontStyleContext,
  LineHeightMapper,
} from './MyText.types';
import { useFontStyles } from './useFontStyles';

const defaultMapToLineHeight: LineHeightMapper = () => undefined;
/**
 * MyTextContext is a React Context that provides the font styles to
 * all nested MyText components.
 */
const MyTextContext = createContext<FontStyleContext>({
  mapToFontFamily: mapToGoogleFontKey,
  mapToLineHeight: defaultMapToLineHeight,
});

export type MyTextProps = TextProps & {
  /**
   * An optional mapper that would get the final font family given the logical font family, font weight, and font style.
   * This defaults to {@link mapToGoogleFontKey}
   */
  mapToFontFamily?: FontFamilyMapper;

  /**
   * An optional mapper that would take the fontSize and translate it to a lineHeight value to ensure consistency across platforms.
   * This defaults to always return `undefined` which lets React Native handle it natively.
   */
  mapToLineHeight?: LineHeightMapper;
};

/**
 * A factory function that creates a custom Text component that inherits font styles
 * (fontFamily, fontStyle, fontWeight) from its parent MyText components unless explicitly overridden.
 *
 * @param defaultMapToFontFamily - The default function to map the font family.
 * @returns A React component that renders a Text element with inherited font styles.
 */
const createMyTextComponent = (defaultMapToFontFamily: FontFamilyMapper) =>
  // eslint-disable-next-line react/display-name
  forwardRef<Text, MyTextProps>(
    ({ children, mapToFontFamily, mapToLineHeight, style, ...props }, ref) => {
      const parentStyles = useContext(MyTextContext);
      const { currentFontStyles, combinedStyle } = useFontStyles(
        mapToFontFamily ?? defaultMapToFontFamily,
        mapToLineHeight ?? defaultMapToLineHeight,
        style,
        parentStyles,
      );

      return (
        <MyTextContext.Provider value={currentFontStyles}>
          <Text ref={ref} style={combinedStyle} {...props}>
            {children}
          </Text>
        </MyTextContext.Provider>
      );
    },
  );

/**
 * MyText is a custom Text component that inherits font styles
 * (fontFamily, fontStyle, fontWeight) from its parent MyText components
 * unless explicitly overridden. It presumes the use of Expo Google Fonts
 * and the fonts are loaded using the `useExpoGoogleFonts` hook.
 */
export const MyText = createMyTextComponent(mapToGoogleFontKey);

/**
 * MyTextE is a custom Text component similar to MyText, but it uses
 * embedded font families instead of Google Fonts by default.
 */
export const MyTextE = createMyTextComponent(mapToEmbeddedFontFamily);
