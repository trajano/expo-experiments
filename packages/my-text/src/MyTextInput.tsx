import { createMyTextInputComponent } from './createMyTextInputComponent';
import {
  mapToEmbeddedFontFamily,
  mapToGoogleFontKey,
} from './mapToEmbeddedFontFamily';

/**
 * MyTextInput is a custom TextInput component that inherits font styles
 * (fontFamily, fontStyle, fontWeight) from its parent MyText components
 * unless explicitly overridden. It presumes the use of Expo Google Fonts
 * and the fonts are loaded using the `useExpoGoogleFonts` hook.
 */
export const MyTextInput = createMyTextInputComponent(mapToGoogleFontKey);

/**
 * MyTextInputE is a custom TextInput component similar to MyTextInput, but it uses
 * embedded font families instead of Google Fonts by default.
 */
export const MyTextInputE = createMyTextInputComponent(mapToEmbeddedFontFamily);
