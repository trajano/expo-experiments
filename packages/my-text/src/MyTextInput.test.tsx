import { render } from '@testing-library/react-native';
import { StyleSheet } from 'react-native';
import { MyText } from './MyText';
import { MyTextInput } from './MyTextInput';

describe('MyTextE', () => {
  it('should render with embedded font family styles on iOS', () => {
    const { getByTestId, toJSON } = render(
      <>
        <MyText style={{ color: 'white' }}>
          <MyText>Email:</MyText>
        </MyText>
        <MyTextInput testID="input" style={styles.input} />
      </>,
    );

    const inputField = getByTestId('input');

    expect(inputField).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });
});
const styles = StyleSheet.create({
  input: {
    borderColor: '#ccc',
    borderWidth: 1,
    padding: 10,
    borderRadius: 5,
    marginTop: 5,
  },
});
