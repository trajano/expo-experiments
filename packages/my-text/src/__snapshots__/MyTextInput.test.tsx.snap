// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyTextE should render with embedded font family styles on iOS 1`] = `
[
  <Text
    style={
      {
        "color": "white",
        "fontFamily": undefined,
        "fontStyle": undefined,
        "fontWeight": undefined,
        "lineHeight": undefined,
      }
    }
  >
    <Text
      style={
        {
          "fontFamily": undefined,
          "fontStyle": undefined,
          "fontWeight": undefined,
          "lineHeight": undefined,
        }
      }
    >
      Email:
    </Text>
  </Text>,
  <TextInput
    style={
      {
        "borderColor": "#ccc",
        "borderRadius": 5,
        "borderWidth": 1,
        "fontFamily": undefined,
        "fontStyle": undefined,
        "fontWeight": undefined,
        "lineHeight": undefined,
        "marginTop": 5,
        "padding": 10,
      }
    }
    testID="input"
  />,
]
`;
